src/mechanics/dofs/DofMatrixContainer.h

```diff
@@ -1,55 +1,98 @@
+    DofMatrixContainer& operator+=(const DofMatrixContainer& rhs)
     {
-        throw Exception(__PRETTY_FUNCTION__, "This method is for vectors only, not for matrices.");
+        for (auto& entry : rhs.mData)
+        {
+            if (mData.find(entry.first) != mData.end())
+                mData[entry.first] += entry.second;
+            else
+                mData[entry.first] = entry.second;
+        }
+        return *this;
+    }
+
+    DofMatrixContainer& operator*=(double scalar)
+    {
+        for (auto& entry : mData)
+            entry.second *= scalar;
+        return *this;
+    }
+
+    friend DofMatrixContainer operator+(DofMatrixContainer lhs, const DofMatrixContainer& rhs)
+    {
+        lhs += rhs;
+        return lhs;
+    }
+
+    friend DofMatrixContainer operator*(DofMatrixContainer lhs, double scalar)
+    {
+        lhs *= scalar;
+        return lhs;
+    }
+
+    //! @brief calculates (*this) += `rhs` * `scalar`
+    //! @remark This is the common case in the numerical integration - summing up all integration point contributions
+    //! scaled with DetJ and the integration point weight.
+    void AddScaled(const DofMatrixContainer& rhs, double scalar)
+    {
+        for (auto& entry : rhs.mData)
+        {
+            if (mData.find(entry.first) != mData.end())
+                mData[entry.first] += entry.second * scalar;
+            else
+                mData[entry.first] = entry.second * scalar;
+        }
     }
 
     friend std::ostream& operator<<(std::ostream& out, const DofMatrixContainer<T>& dofMatrix)
     {
-        for (size_t i = 0; i < dofMatrix.mData.size(); ++i)
+        for (auto& entry : dofMatrix.mData)
         {
-            auto xy = CantorPairingFunctionReverse(i);
-            out << "=== " << xy.first << " " << xy.second << " ===" << std::endl;
-            out << dofMatrix.mData[i] << std::endl;
+            auto& dofs = entry.first;
+            out << "=== " << dofs.first.GetName() << " " << dofs.second.GetName() << " ===" << std::endl;
+            out << entry.second << std::endl;
         }
         out << "====" << std::endl;
         return out;
     }
 
 private:
-    //! @brief https://en.wikipedia.org/wiki/Pairing_function#Cantor_pairing_function
-    //! maps NxN --> N uniquely, deterministic, nice.
-    static int CantorParingFunction(int k1, int k2)
+    using DofPair = std::pair<DofType, DofType>;
+    struct CompareDofPairs
     {
-        return k2 + (k1 + k2) * (k1 + k2 + 1) / 2;
-    }
-
-    static std::pair<int, int> CantorPairingFunctionReverse(int z)
-    {
-        int w = std::floor((std::sqrt(8 * z + 1) - 1) / 2);
-        int t = (w * w + w) / 2;
-        return std::make_pair(z - t, w - z + t);
-    }
+        bool operator()(const DofPair& a, const DofPair& b) const
+        {
+            // check for equiv(a.first, b.first)
+            if (!(a.first.GetName() < b.first.GetName()) && !(b.first.GetName() < a.first.GetName()))
```

Comment:
`DofType` has a unique `.Id()` but not necessarily a unique `.GetName()`.

Reply:
Yes I already realized that while cycling home 😬